engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -85,79 +91,119 @@ class EngineJobExecutionActor(jobKey: BackendJobDescriptorKey,
       }
     case Event(response: BackendJobPreparationFailed, NoData) =>
       context.parent forward response
       context stop self
       stay()
   }
 
-  // When CheckingCallCache, the FSM always has EJEAJobDescriptorData
+  private def addHashesAndStay(data: NoResponseData, hashes: CallCacheHashes): State = {
+    val updatedData = data.copy(hashes = Option(Success(hashes)))
+    stay using updatedData
+  }
+
   private val callCachingReadResultMetadataKey = "Call caching read result"
   when(CheckingCallCache) {
-    case Event(HashError(t), JobDescriptorData(jobDescriptor, bjeaProps)) =>
+    case Event(HashError(t), data: NoResponseData) =>
       writeToMetadata(Map(callCachingReadResultMetadataKey -> s"Hashing Error: ${t.getMessage}"))
       recordHashError(t)
-      runJob(jobDescriptor, bjeaProps)
-    case Event(CacheMiss, JobDescriptorData(jobDescriptor, bjeaProps)) =>
+      runJob(data)
+    case Event(CacheMiss, data: NoResponseData) =>
       writeToMetadata(Map(callCachingReadResultMetadataKey -> "Cache Miss"))
-      log.info(s"Cache miss for job ${jobDescriptor.key.call.fullyQualifiedName}, index ${jobDescriptor.key.index}")
-      runJob(jobDescriptor, bjeaProps)
-    case Event(CacheHit(cacheResultId), JobDescriptorData(jobDescriptor, _)) =>
-      writeToMetadata(Map(callCachingReadResultMetadataKey -> s"Cache Hit (from result ID $cacheResultId)"))
-      log.info(s"Cache hit for job ${jobDescriptor.key.call.fullyQualifiedName}, index ${jobDescriptor.key.index}! Copying cache result $cacheResultId")
-      lookupCachedResult(jobDescriptor, cacheResultId)
+      log.info("Cache miss for job {}", jobTag)
+      runJob(data)
+    case Event(hit: CacheHit, data: NoResponseData) =>
+      writeToMetadata(Map(callCachingReadResultMetadataKey -> s"Cache Hit (from result ID ${hit.cacheResultId})"))
+      log.info("Cache hit for {}! Fetching cached result {}", jobTag, hit.cacheResultId)
+      fetchCachedResults(data, jobKey.call.task.outputs, hit)
+    case Event(hashes: CallCacheHashes, data: NoResponseData) =>
+      addHashesAndStay(data, hashes)
+  }
+
+  when(FetchingCachedOutputsFromDatabase) {
+    case Event(CachedOutputLookupSucceeded(cachedJobOutputs, cacheHit), data: NoResponseData) =>
+      copyCacheHit(cacheHit, cachedJobOutputs, data)
+    case Event(CachedOutputLookupFailed(metaInfoId, error), data: NoResponseData) =>
+      log.info("Can't make a copy of the cached job outputs for {} due to {}. Running job.", jobTag, error)
+      runJob(data)
+    case Event(hashes: CallCacheHashes, data: NoResponseData) =>
+      addHashesAndStay(data, hashes)
+  }
+
+  when(BackendIsCopyingCachedOutputs) {
+    case Event(response: SucceededResponse, data @ NoResponseData(_, _, Some(Success(hashes)))) if effectiveCallCachingMode.writeToCache =>
+      saveCacheResults(hashes, response.updatedData(data))
+    case Event(response: SucceededResponse, data @ NoResponseData(_, _, Some(Success(hashes)))) =>
```

Comment:
In other words, this `when` should catch the same variety of situations as the `when(RunningJob)`